teps/0130-pipeline-level-build-cache.md

```diff
@@ -0,0 +1,557 @@
+proposal will be implemented, this is the place to discuss them.
+
+If it's helpful to include workflow diagrams or any other related images,
+add them under "/teps/images/". It's upto the TEP author to choose the name
+of the file, but general guidance is to include at least TEP number in the
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+This needs to be fleshed out in more detail, but the sidecar would likely be implemented as a deployment with one pod plus a service.
+If the sidecar has readiness or startup probes, the PipelineRun would wait for them to succeed before starting the Pipeline Tasks.
+We could use a [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+to prevent other PipelineRuns from being able to use the sidecar.
+
+## Design Evaluation
+
+### Reusability
+
+A few sidecar configurations, such as docker daemons and database containers, likely make up the majority of use cases.
+It's reasonable to ask whether we'd want to introduce a concept of a reusable sidecar.
+For example, we could include a docker daemon sidecar in the catalog,
+and allow Pipelines to contain inline sidecar specifications or references to sidecars
+(stored on the hub, on the cluster, in git, etc, just like Tasks).
+
+However, the lifespan of a reusable sidecar wouldn't be tied to the lifespan of a PipelineRun, and managing it shouldn't be the responsibility of the PipelineRun controller.
+This proposal is only scoped to sidecars defined inline in a Pipeline.
+
+### Simplicity
+
+The current user experience isn't flexible enough for many common docker-related use cases involving more complex logic than a build and push.
+
+### Flexibility
+
+Allowing users to specify Pipeline-level sidecars is more flexible than first class support for building images,
+as discussed in the alternative ["First-class docker support"](#first-class-docker-support).
+
+### Conformance
+
+Sidecar configuration already contains some Kubernetes-specific configuration,
+such as liveness/readiness probes. This proposal doesn't introduce new Kubernetes-specific
+configuration, but it allows it to be specified in Pipelines as well as Tasks.
+
+### User Experience
+
+Without a Pipeline-level sidecar, users would have to write a first Task that creates a service and a deployment,
+and a finally Task that tears it down.
+
+This Task would have to use a service account with permissions to create and delete services and deployments,
+and mount a Workspace with k8s credentials.
+
+The user would also have to write a script to create the service and deployment, wait for it to start up,
+and produce the service address as a result. (They probably wouldn't want to hard-code the service address,
+because it would prevent multiple instances of the PipelineRun from running in the same namespace at the same time,
+but they could create a service with a name prefixed by the PipelineRun name.)
```

Comment:
All this is technically *true*, but it's also why we try to make `Task` re-usable items. Ideally, creating a docker daemon and killing at the end is the same everywhere (and can be parametrized enough to be re-used 99% of the time). Assuming this, these sentence change into "users would have to use a given Task, …", which.. is slightly different.

Reply:
That's fair, I've rephrased